packages/graalvm/src/main/kotlin/elide/runtime/node/querystring/NodeQuerystring.kt

```diff
@@ -12,36 +12,356 @@
+        2 -> decode(args[0], args[1], null, null)
+        3 -> decode(args[0], args[1], args[2], null)
+        4 -> decode(args[0], args[1], args[2], args[3])
+        else -> throw JsError.typeError("Invalid number of arguments to `querystring.decode`")
+      }
+    }
+
+    QS_ENCODE -> ProxyExecutable { args ->
+      when (args.size) {
+        1 -> encode(args[0], null, null, null)
+        2 -> encode(args[0], args[1], null, null)
+        3 -> encode(args[0], args[1], args[2], null)
+        4 -> encode(args[0], args[1], args[2], args[3])
+        else -> throw JsError.typeError("Invalid number of arguments to `querystring.encode`")
+      }
+    }
+
+    QS_ESCAPE -> ProxyExecutable { args ->
+      when (args.size) {
+        1 -> escape(args[0])
+        else -> throw JsError.typeError("Invalid number of arguments to `querystring.escape`")
+      }
+    }
+
+    QS_PARSE -> ProxyExecutable { args ->
+      when (args.size) {
+        1 -> parse(args[0], null, null, null)
+        2 -> parse(args[0], args[1], null, null)
+        3 -> parse(args[0], args[1], args[2], null)
+        4 -> parse(args[0], args[1], args[2], args[3])
+        else -> throw JsError.typeError("Invalid number of arguments to `querystring.parse`")
+      }
+    }
+
+    QS_STRINGIFY -> ProxyExecutable { args ->
+      when (args.size) {
+        1 -> stringify(args[0], null, null, null)
+        2 -> stringify(args[0], args[1], null, null)
+        3 -> stringify(args[0], args[1], args[2], null)
+        4 -> stringify(args[0], args[1], args[2], args[3])
+        else -> throw JsError.typeError("Invalid number of arguments to `querystring.stringify`")
+      }
+    }
+
+    QS_UNESCAPE -> ProxyExecutable { args ->
+      when (args.size) {
+        1 -> unescape(args[0])
+        else -> throw JsError.typeError("Invalid number of arguments to `querystring.unescape`")
+      }
+    }
+
+    else -> null
+  }
+
+  override fun decode(
+    str: Value,
+    sep: Value?,
+    eq: Value?,
+    options: Value?
+  ): QueryParams = parse(str, sep, eq, options)
+
+  override fun encode(
+    obj: Value,
+    sep: Value?,
+    eq: Value?,
+    options: Value?
+  ): String = stringify(obj, sep, eq, options)
+
+  override fun escape(str: Value): String {
+    val bytes = valueToString(str).toByteArray(StandardCharsets.UTF_8)
+    return codec.encode(bytes).toString(Charsets.UTF_8)
+  }
+
+  override fun parse(
+    str: Value,
+    sep: Value?,
+    eq: Value?,
+    options: Value?
+  ): QueryParams {
+
+
+    val string = valueToString(str)
+    val separator = sep?.let(::valueToString) ?: "&"
+    val equals = eq?.let(::valueToString) ?: "="
+    val parseOptions = ParseOptions.fromGuest(options)
 
-  override fun getMemberKeys(): Array<String> = emptyArray()
-  override fun getMember(key: String?): Any? = null
+    val result = mutableMapOf<String, Any>()
+
+    if (string.isBlank()) {
+      return of(result)
+    }
+
+    var keyCount = 0
+    string.split(separator)
```

Comment:
prefer using `String.splitToSequence` here instead, as it yields values progressively instead of processing the whole string up front